File: server/frontend/src/shared/api/protectedSpace.ts
import { apiRequest, USE_MOCKS } from "./client";
import { clearReportsCache } from "./reports";
import { setVaultToken } from "./vaultToken";
import type {
  ProtectedSpaceRead,
  ProtectedSpaceStatus,
  ProtectedSpaceUnlockResponse,
} from "../types/protectedSpace";

let mockProtectedSpace: ProtectedSpaceRead | null = null;
let mockUnlockExpiresAt: string | null = null;

export function clearProtectedVaultToken(): void {
  setVaultToken(null);
  mockUnlockExpiresAt = null;
  clearReportsCache();
}

export async function getProtectedSpaceStatus(): Promise<ProtectedSpaceStatus> {
  if (USE_MOCKS) {
    return {
      exists: mockProtectedSpace !== null,
      workspace_id: mockProtectedSpace?.workspace_id ?? null,
      is_unlocked: mockUnlockExpiresAt !== null && new Date(mockUnlockExpiresAt).getTime() > Date.now(),
      expires_at: mockUnlockExpiresAt,
    };
  }

  return apiRequest<ProtectedSpaceStatus>("/api/v1/protected-space/status");
}

export async function createProtectedSpace(password: string): Promise<ProtectedSpaceRead> {
  if (USE_MOCKS) {
    if (mockProtectedSpace) {
      throw new Error("Защищённое пространство уже создано");
    }
    mockProtectedSpace = {
      workspace_id: 999,
      name: "Защищённое пространство 🔒",
      is_enabled: true,
      created_at: new Date().toISOString(),
    };
    return mockProtectedSpace;
  }

  return apiRequest<ProtectedSpaceRead>("/api/v1/protected-space", {
    method: "POST",
    body: JSON.stringify({ password }),
  });
}

export async function unlockProtectedSpace(password: string): Promise<ProtectedSpaceUnlockResponse> {
  if (USE_MOCKS) {
    const expiresAt = new Date(Date.now() + 10 * 60 * 1000).toISOString();
    const response = {
      workspace_id: mockProtectedSpace?.workspace_id ?? 999,
      vault_token: `mock-vault-token-${Date.now()}`,
      expires_at: expiresAt,
    };
    setVaultToken(response.vault_token);
    mockUnlockExpiresAt = expiresAt;
    return response;
  }

  const response = await apiRequest<ProtectedSpaceUnlockResponse>("/api/v1/protected-space/unlock", {
    method: "POST",
    body: JSON.stringify({ password }),
  });
  setVaultToken(response.vault_token);
  return response;
}

export async function lockProtectedSpace(): Promise<void> {
  try {
    if (!USE_MOCKS) {
      await apiRequest<void>("/api/v1/protected-space/lock", { method: "POST" });
    }
  } finally {
    clearProtectedVaultToken();
  }
}

export async function changeProtectedSpacePassword(
  currentPassword: string,
  newPassword: string,
): Promise<void> {
  await apiRequest<{ detail: string }>("/api/v1/protected-space/change-password", {
    method: "POST",
    body: JSON.stringify({
      current_password: currentPassword,
      new_password: newPassword,
    }),
  });
  clearProtectedVaultToken();
}

File: server/Backend/src/schemas/protected_space.py
from datetime import datetime

from pydantic import BaseModel, Field, field_validator


class ProtectedSpacePasswordMixin(BaseModel):
    @field_validator("password", "new_password", check_fields=False)
    @classmethod
    def strip_password_edges(cls, value: str) -> str:
        value = value.strip()
        if len(value) < 12:
            raise ValueError("Защитный пароль должен быть не короче 12 символов")
        return value


class ProtectedSpaceCreate(ProtectedSpacePasswordMixin):
    password: str = Field(min_length=12, max_length=1024)


class ProtectedSpaceUnlock(BaseModel):
    password: str = Field(min_length=1, max_length=1024)


class ProtectedSpaceChangePassword(ProtectedSpacePasswordMixin):
    current_password: str = Field(min_length=1, max_length=1024)
    new_password: str = Field(min_length=12, max_length=1024)


class ProtectedSpaceStatus(BaseModel):
    exists: bool
    workspace_id: int | None = None
    is_unlocked: bool = False
    expires_at: datetime | None = None


class ProtectedSpaceRead(BaseModel):
    workspace_id: int
    name: str
    is_enabled: bool
    created_at: datetime


class ProtectedSpaceUnlockResponse(BaseModel):
    workspace_id: int
    vault_token: str
    expires_at: datetime


class ProtectedSpaceMessage(BaseModel):
    detail: str
